[Serialization] Drop overriding relationship in constructors when safe

xymus opened the pull request:
Deserialization recovery lead the compiler to drop public constructors overriding internal constructors. This PR limits the logic to dropping only the overriding relationship instead of the whole constructor. This applies when the overriden constructor fails to deserialize and only when the overriding relationship was marked as not affecting ABI. This should realign the behavior between swiftmodules and swiftinterfaces.

rdar://104704832

Fixes: https://github.com/apple/swift/issues/63242

This is the equivalent of https://github.com/apple/swift/pull/63068 for constructors.

xymus committed:
[Serialization] Drop overridden relationship in constructors when safe

Deserialization recovery lead the compiler to drop public constructors
overridding internal constructors. This limits the logic to dropping the
overriding relationship instead of the whole constructor. This applies
when the overriden constructor fails to deserialize and only when the
overriding relationship was marked as not affecting ABI.

rdar://104704832

xymus commented:
@swift-ci Please smoke test

bnbarham commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3130,15 +3131,23 @@ class DeclDeserializer {
+    if (overriddenOrError) {
+      overridden = overriddenOrError.get();
+    } else {
+      if (overriddenAffectsABI || !ctx.LangOpts.EnableDeserializationRecovery) {
```
Could we set `overridden = nullptr` if we're allowing compiler errors in this case (ie. `MF.allowCompilerErrors`)?

xymus replied:
Sure 👍

bnbarham commented on lib/Serialization/Serialization.cpp:
```diff
@@ -4494,9 +4494,12 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
       if (firstTimeRequired && overridden->isRequired())
         firstTimeRequired = false;
+    }
+    bool overriddenAffectsABI = overriddenDeclAffectsABI(ctor, overridden);
```
Is the relationship actually useful if it doesn't?

xymus replied:
That's a good question. For non-resilient modules it could be used by optimizations. For resilient modules in most cases it should now be dropped by deserialization safety for the public-override-internal case, and the clang case is there more as tolerance for poor modularization and such I believe. I think we could try always dropping it when reading from a resilient module.

xymus committed:
[Serialization] Limit ignoring internal overriden decl to lib evolution

At deserialization, we can drop the overriding relationship if the
overriden decl is internal and only when the overrider is in a resilient
module.

xymus committed:
[Serialization] Drop any failed overriding relationship when allowing errors

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
https://github.com/apple/swift-driver/pull/1281

@swift-ci please smoke test macOS
